### PR TITLE
Refactor hostedServices into background jobs

### PR DESCRIPTION
**Pull request by andrewmckaskill**
**DRAFT PR for Discussion**

### Prerequisites

- [ ] I have added steps to test this contribution in the description below

### Description

This PR refactors the current jobs that are run in the background to split the task from the engine that runs them.

This should allow for the following enhancements in the future:
- support alternate engines in the future (such as Hangfire, Quartz)
- formalising contract for an "Umbraco" background job (eg. ServerRole, Period, MainDom)
- easier to write background jobs
- easier for developers to add new recurring jobs and make sure that the jobs run when / where they are supposed to
- easier to add database logging of task failures to produce dashboard

### Details

There is a new `IReccuringBackgroundJob` interface to describe the contract of a background job. These get injected into the DI container separately from the engine that is running them.

The "engine" for right now is still a .net core `HostedService` - but I've split the implementation up a little to make it easier to test and to split the responsibility.

The `RecurringBackgroundJobHostedService<Job>` class is a generic class that takes a IRecurringJob as a parameter. I used generics here so that if you wanted to you could inject this class directly into the DI container and it would run. If (for deprecation reasons) we have to keep the existing jobs that extend from HostedService and are injected directly, then I would swap them over to extend from this class instead. The functionality isn't 100% the same, but the lifecycle and usage pattern would match.
This class manages the "Umbraco" requirements for the RecurringBackgroundJob (eg. ServerRole, MainDom). It extends from `RecurringHostedServiceBase` class, which manages the requirements for running a task on a periodic basis. I like the split between the two as it keeps the classes small and easier to understand.
 
I also created a `RecurringBackgroundJobHostedServiceRunner` class. This class IS a hosted service, and it's job is to identify any `IRecurringBackgroundJob` jobs in the DI container and spin up HostedServices for them. I've done it this way for two reasons:
First, we avoid injecting HostedServices for each job directly in the DI. Using a different engine in the future would require you to un-register those HostedServices first which isn't great.
Second, we can also gather up any `IRecurringBackgroundJob` jobs that have been added to the DI container from user code.

### Still to do

**Short-term**
- [ ] Unit tests for new components
- [ ] Deprecating / Removing / fixing the old HostedService background jobs
- [ ] Documentation

**Long Term**
- [ ] Database table for storing task name / servername / last run date / last run status
- [ ] Simple backoffice UI to surface task status

### CODE REVIEW

**Comment by JasonElkin** [comment]
Hey @andrewmckaskill, thanks for contributing and making an awesome start on this 🙌. On a personal note it's certainly an enhancement I'd like to see!

As it's still a WIP I'll convert to to a draft PR for now. Once It's ready for the limelight, convert it back and it should set off the sirens for one of us core collabs to come and do the last round of testing and merge it in!
 
In the meantime, I'll flag this to HQ so that we can get their input on it. 

I know scheduled tasks are part of Umbraco that people in the community have had mixed experiences with, so it's probably worth sharing this PR and striking up a conversation on Discord, Twitter etc. to see if there are any other insights from the community that might be helpful.

Thanks again!
Jas

**Comment by andrewmckaskill** [comment]
@bergmania over to you for review.
Outstanding questions:
- Should the Jobs be moved into the Core project?
- What description do you want in the Obsolete attributes
- Should we consider adding a "ServerNode" property to limit running the job on a specific server node (or do we leave this for later)

File: src/Umbraco.Infrastructure/BackgroundJobs/IRecurringBackgroundJob.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+/// <summary>
+///     A recurring background job
+/// </summary>
+public interface IRecurringBackgroundJob
+{
+    static readonly TimeSpan DefaultDelay = System.TimeSpan.FromMinutes(3);
+    static readonly ServerRole[] DefaultServerRoles = new[] { ServerRole.Single, ServerRole.SchedulingPublisher }; 
+
+    /// <param name="period">Timespan representing how often the task should recur.</param>
+    /// <param name="delay">
```
Comment (bergmania) [inline src/Umbraco.Infrastructure/BackgroundJobs/IRecurringBackgroundJob.cs:14]: These seems out of place

**Comment by bergmania** [comment]
Hi Andrew.. Sorry for the very late response..

I finally found time to look into the PR - It's great stuff!

To answer your questions:
>Should the Jobs be moved into the Core project?
Not all jobs can be moved anyway due to some dependencies, so I think it's okay to leave them in infrastructure.

>What description do you want in the Obsolete attributes
The current descriptions are great, we just need to add that we expect that API to be removed in Umbraco 14.

>Should we consider adding a "ServerNode" property to limit running the job on a specific server node (or do we leave this for later)
I think we should just leave it for now.

File: src/Umbraco.Infrastructure/BackgroundJobs/IRecurringBackgroundJob.cs (hunk 1 above)
Comment (andrewmckaskill) [reply src/Umbraco.Infrastructure/BackgroundJobs/IRecurringBackgroundJob.cs:16]: Good catch! I've updated those now.

**Comment by andrewmckaskill** [comment]
@bergmania - All points raised should be resolved now. Any update on when it's likely to get merged? 
Once it's in I'll take a look at updating the main documentation.

**Comment by bergmania** [comment]
Great work @andrewmckaskill.. Thank you 💪

**Pull request merged**